Detsim electron simulation functions

gondiaz opened the pull request:
Following issue #691:
4. upload functions to simulate electron drift (diffusion/attachment etc) in invisible_cities/cities/detsim_simulate_electrons.py

This module contains three functions, that generate, drift and diffuse the secondary electrons produced by the energy hits.

andLaing commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,52 @@
+import numpy  as np
+
+from typing    import   List, Tuple, Callable
+
```
Again, I prefer a line per import

andLaing commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,52 @@
+#######################################
+def generate_electrons(energies    : np.array,
+                       wi          : float,
+                       fano_factor : float ) -> np.array:
```
Bit of alignment here maybe

andLaing commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,52 @@
+                       wi          : float,
+                       fano_factor : float ) -> np.array:
+    """ generate secondary electrons from energy deposits
+    """
```
More info

gonzaponte replied:
@andLaing is this comment outdated or do you still require more info to be added?

andLaing replied:
I just noticed this question, sorry for the delay in replying.

I think it explains things better and makes the choice clearer. I think the English could be improved but that's not urgent, we can do a quick edit before merging.

andLaing commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,52 @@
+    nes  = np.array(energies/wi, dtype = int)
+    pois = nes < 10
+    nes[ pois] = np.random.poisson(nes[pois])
+    nes[~pois] = np.round(np.random.normal(nes[~pois], np.sqrt(nes[~pois] * fano_factor)))
```
I've never been fully convinced of the necessity of this differentiation. Maybe @paolafer or @gonzaponte have more insight. In any case, I think the `< 10` needs to be justified to some extent. Do you have a notebook or similar with plots? What difference do we get in distribution if we always use the fano factor? What change in performance?

gonzaponte replied:
I think this comes from C++. Numpy handles different values of the mean of the distribution automatically. I haven't checked the performance, though.

P.S: The value chosen by numpy is also 10.

andLaing replied:
You mean that if you use a mean value greater than 10 with `np.random.poisson` it really uses `np.random.normal`? So in our case the only real difference if we just used normal is that the standard deviation will be scaled by a factor of `sqrt(fano_factor)` for signals below 10 pe?

gonzaponte replied:
No. For values of the mean greater than 10 it uses a different algorithm, but it would still be a Poisson distribution.

But I had forgotten about the fano factor.  I was thinking this in the wrong way (assuming we used Gauss just as a limit of the Poisson for large values of the mean). We actually want to generate a normal distribution with a smaller standard deviation than the one coming from the Poisson limit. Therefore we *only* generate Poisson when we cannot rely on the gaussian distribution. The problem with the Poisson distribution is that the std deviation and the mean are not independent and we cannot make it narrower as we do with the gaussian.

So in conclusion, hits with less than 10 ie will have a wider distribution than they should.

gondiaz replied:
Can't we just simply add the `fano_factor` in the poisson?

gonzaponte replied:
No, that was precisely my point when I said _"with the Poisson distribution [...] the std deviation and the mean are not independent"_

jmalbos replied:
I don't know what's the statistical limit of the Fano factor. Does it work the same way for a population of just a few electrons? I wouldn't be surprised if it's never been measured for xenon gas at energies below 1 keV. 
In any case, if the product _NF_ is < 1 (which is the case when _N_<10), the resulting Poisson distribution yields most of the time either 0 or 1 electrons. The idea was to avoid that behavior, given that the _W_ energy is just an effective average. In summary, I'd leave it that way for now.

andLaing replied:
OK, so I think we probably all agree to leave it as is? But I'd add a comment with a bit of explanation of the choice, otherwise it's a bit confusing.

gonzaponte replied:
Coming back to this, please add a short inline comment summarizing the logic behind it so we don't forget.

gondiaz replied:
Following @jmalbos comment, the new implementation decides whether to "poissonize" or "normalizate" based on the variance NF<1. I think this criteria makes the function less confusing than the 10 electrons "dont't know where it comes" value.

andLaing commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,52 @@
+#######################################
+######### ELECTRON SIMULATION #########
+#######################################
+def generate_electrons(energies    : np.array,
```
I'd maybe rename to `generate_ionisation_electrons` (or spelling with z, I think US standard is IC standard normally) just to be clear.

andLaing commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,52 @@
+    ts  = zs / drift_velocity
+    nes = electrons - np.random.poisson(electrons * (1. - np.exp(-ts/lifetime)))
+    nes[nes < 0] = 0
+    return nes
```
This function doesn't have the functionality necessary to use a lifetime map as is the standard at the moment. Or am I missing something? I think it needs to be generalised.

gondiaz replied:
Thats true, but I think this could be better implemented once the PR #698 is approved, since the way of implementing this callable (that is, a `lifetime(x, y)` function) would be the same as for the psf functions. Once approved, this function would be easily modified.

jahernando replied:
Including the dependence of lifetime with (x, y) can be done in a future upgrade of the detsim code.

andLaing commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,52 @@
+                      electrons              : np.array,
+                      transversal_diffusion  : float,
+                      longitudinal_diffusion : float)\
+                      -> Tuple[np.array, np.array, np.array, np.array]:
```
Be careful of the number of returns if you want to be explicit

andLaing commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,52 @@
+                      ys                     : np.array,
+                      zs                     : np.array,
+                      electrons              : np.array,
+                      transversal_diffusion  : float,
```
transverse_diffusion

transversal means something else

andLaing commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,78 @@
+    longitudinal_diffusion   = conf.get("longitudinal_diffusion")
+
+    return wi, fano_factor, lifetime, drift_velocity,\
+           transversal_diffusion, longitudinal_diffusion
```
There are a lot of fixtures here for something that could probably be one and just have the values written into them rather than read from a config.

I'd think a bit more about the scope of the information you need for your test (and maybe others) and write it into a fixture of similar format to those already in IC. If you think it's something that could be useful in its entirety or partially in other tests in other files, we can put it in a conftest.py file and set the scope accordingly, if you think it's just for this module put it here and set the scope to the module.

andLaing commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,78 @@
+    (dxs, dys, dzs) = diffuse_electrons(xs, ys, zs, electrons,
+                                        transversal_diffusion,
+                                        longitudinal_diffusion)
+    assert len(dxs)==len(dys)==len(dzs)==np.sum(electrons)
```
While it's quite nice to have a test which combines the three functions we need tests for each individual one too to make sure the functionality is there.

Also, we probably need to think of something a bit more robust than just checking the lengths of the outputs.

andLaing reviewed (dismissed):
General comments:
The files are in the wrong place. `cities` should contain cities and components of source type (in components.py). These are general functions that belong either in a folder dedicated to detector simulation or in reco maybe. I think they should be moved.

jmunozv commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,52 @@
+    dys  = np.random.normal(ys, sqrtz * transversal_diffusion)
+    dzs  = np.random.normal(zs, sqrtz * longitudinal_diffusion)
+
+    return (dxs, dys, dzs)
```
If I fully understand the code, right now we are assuming that every ionization e- is being created at the same time.
This is always true except for Bi214 simulations. The decay of Bi214 lead to hits "delayed" tens or hundreds of microseconds with respect the first energy depositions. This time difference must be considered when diffusing electrons in order to properly reconstruct their Z positions and therefore, the detectors response.

andLaing replied:
In fact I can think of other types of event where this is relevant too (neutron captures for example). Things like this are included in full sim but we'll have to be careful to have this functionality here too. Things like this were one of the to-dos left hanging in the old versions

gondiaz replied:
That's right, but if you look at dataflow PR #700, you see that the part that manages the buffer times is different from the one from electron simulation. I think that it would be more natural to include this shift in buffer times directly there, since here we just implement the effects of attachment and diffusion, which is independent of relative times between hits. Nevertheless, we could add this as a PR once detsim is approved.

andLaing replied:
We're not talking about buffer time but absolute time. The MC/hit has the time (in PR #694 you don't even return the info which should be sorted) as well as position, in the BiPo events Javi mentioned the hits could be at ver similar x, y, z position but separated by multiple µs or even ms.

gondiaz replied:
Yes, we are saying the same. Sure, PR #694, must have this info too. What I say is that if you look at PR #700, the "buffer time" part also computes the drift times (`z/drif_velocity`). So to take into account the absolute times (or relative times between hit at initial time 0), this is were this should be implemented, by summing up the drift time plus the hit production time.

andLaing replied:
I don't think I agree, especially if we want to be as flexible as possible but, anyway, let's see how things progress with the other functions

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+def generate_ionization_electrons(wi          : float,
+                                  fano_factor : float,
+                                  energies    : np.ndarray) -> np.ndarray:
+    """ generate ionization secondary electrons from energy deposits
```
```suggestion
    """ generate secondary ionization electrons from energy deposits
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+                                  energies    : np.ndarray) -> np.ndarray:
+    """ generate ionization secondary electrons from energy deposits
+
+    Parrameters:
```
```suggestion
    Parameters:
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+
+    Parrameters:
+        :wi: float
+            ionization yield
```
```suggestion
            Mean ionization energy
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+        :nes: np.ndarray
+            the ionization electrons per hit
+    """
+    nes  = np.array(energies/wi, dtype = int)
```
`dtype=int` is wrong. This is the **average** number of ie produced (the mean of the poisson distribution), therefore it can, and must, be a float.

gondiaz replied:
Right, this is a bug that I forgot to fix.
I have removed the `np.round` in the normal distribution and added a protection against negative values, which is very unlikely for smal fano factor, but I think is worth to include.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+        :lifetime: float
+            electron lifetime
+        :drif_velocity: float
+            drif velocity at the active volume of the detector
```
```suggestion
        :drift_velocity: float
            drift velocity at the active volume of the detector
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+        :drif_velocity: float
+            drif velocity at the active volume of the detector
+        :zs:
+            the z hits
```
```suggestion
            z coordinate of the ionization hits
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+            drif velocity at the active volume of the detector
+        :zs:
+            the z hits
+        :electrons:
```
```suggestion
        :electrons:
            Number of ionization electrons of each hit.
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+        :electrons:
+    Returns:
+        :nes: np.ndarray
+            the drifted electrons per initial hit
```
```suggestion
            Number of ionization electrons that reach the EL gap
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+    ts  = zs / drift_velocity
+    nes = electrons - np.random.poisson(electrons * (1. - np.exp(-ts/lifetime)))
+    nes[nes < 0] = 0
+    return nes
```
This is not correct. The correct implementation is
```
    @np.vectorize
    def attachment(n_ie, t):
        return np.count_nonzero(-lifetime * np.log(np.random.uniform(size=n_ie)) > t)
    return attachment(nes, ts)
```
Notice that there is no need to clip at 0.

gondiaz replied:
My understanding is the following:

1. The exponential behaviour of attachment is n'= n*exp(-t/lt), being n' the ie that arrive at the EL from a hit with n ie.
2. Then you define the psf that a single ie reach the EL, that would be: p(t) = exp(-t/lt)/lt. This would be the time distibution of arrival times of ie.
I don't find any physical motivated reason to define this since the ies are associated with a Z position, then an arrival time.
3. Using the cumulative of this distribution (C(t)), you compute a random arrival time (tr) for a single ie as tr=-lt*log(1-C(t)).
4. If the real arrival time (the one from the hit) is less that this tr, then the ie has reached the EL.

This is mathematically correct but it is difficult to me to understand the physical meaning of defining this distribution.

Where is this code coming from?

gonzaponte replied:
It is just the correct way to sample an exponential distribution...
https://en.wikipedia.org/wiki/Exponential_distribution#Generating_exponential_variates

`-lifetime * np.log(np.random.uniform(size=n_ie)) ` generates n_ie numbers that are distributed according to 1/lifetime exp(-t/lifetime). Then we compare those numbers with the drift time, which determines whether or not each ie reached the gate, as you say. `np.count_nonzero` simply counts how many of the electrons reached the gate.

> I don't find any physical motivated reason to define this since the ies are associated with a Z position, then an arrival time.

I don't know what do you mean here. drift_time = z / drift_velocity and the attachment is parameterized as a function of that. It could be parameterized as a function of z, but it is more common (and suitable to compare with data) to do it as a function of drift time.

gondiaz replied:
Yes, this is the method for generating randoms using the "inverse transform sampling" that I mentioned in 3.

Probably I am not explaning myself good enought. I mean that the line of thought is different than I expected.
Imagine a single hit at the active volume, so the pair of values (n, t=Z/dv), being n the number of ionization electrons and t the drift time. We want to compute the number of those electrons reaching the EL (n') from this hit.

-My thought is: just compute n'= n*exp(-t/lt).

-Your thought is a bit different: you generate n exponentially distributed arrival times (tr), and n' would be the number of those times smaller than the hit time (count(tr>t)).

Both ways are mathematically equivalent somehow, but I found yours a bit more confused since it is based in a time-of-arrrival distribution, but the time of each hit is something fixed for each hit.

gonzaponte replied:
The exponential trick is accurate enough for a large number of initial ie. For a small number, it is not. Imagine a small  (5, for instance) number of atoms decaying. The exponential distribution will tell you the *average* number of atoms left after some time t. If you want to simulate the actual number, you need to do something like what I propose. However, if you have 10000 atoms and you don't wait too long, the number of atoms remaining will be large enough for individual decays to dilute in the global trend.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+    """
+    xs = np.repeat(xs, electrons.astype(int))
+    ys = np.repeat(ys, electrons.astype(int))
+    zs = np.repeat(zs, electrons.astype(int))
```
electrons should be of type int before. If it is not, make sure to fix it.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+        :xs, ys, zs: np.ndarray (1D of size: #hits)
+            postions of initial hits
+        :electrons:
+            electrons at the EL before drifting per hit
```
```suggestion
            Number of ionization electrons per hit before drifting
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+            electrons at the EL before drifting per hit
+    Returns:
+        :dxs, dys, dzs: np.ndarray (1D of size: #hits x #total electrons)
+            diffused positions
```
```suggestion
        :dxs, dys, dzs: np.ndarray (1D of size: #hits x #electrons-per-hit)
            Diffused positions at the EL.
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+
+    # substitute z<0 to z=0
+    sel = zs<0
+    zs[sel] = 0
```
Is it possible to have zs<0 from nexus, @paolafer?
This should also be done after applying diffusion, I think.

gondiaz replied:
I added this lines as a temporal fix because some nexus hits were negative (with very smal values of the order of -1e-14). If I am correct, @jmalbos said it was related with the precission of float, so could be dowgraded. Since this was still not done in nexus I kept the fix, which in fact is neccesary, since negative values were entering in a square root.

gonzaponte replied:
Ok. How about adding the same protection after diffusion? Opinions? @jmalbos @paolafer. I don't think it is very relevant, but should they be clipped to 0 or left as they are? My guess is that it will just become an earlier start of the S2 in the waveform.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+             dx         : np.ndarray,
+             dy         : np.ndarray,
+             dz         : np.ndarray)\
+             ->Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
```
I don't understand why we need this function. If it is because we rely on light tables there is a better way to do it, I think. If it is to speed up the calculations, I am not sure it does.

gondiaz replied:
If you look at `pes_at_sipms` function in PR #734, for each hit at the EL we compute the distance between this hit and each of the sensors (actually not all the sensors, but the sensors in a framed region at TP).
So this matrix (`distances`) has shape (nhits, nsensors). Now, for each distance in the matrix, we need to go to the PSF to take the corresponding value. PSF is not a single number, since we have PSFs for several EL partitions, so the PSF values would be a matrix of shape (nhits, nsensors, npartitions). 
This matrix becames huge for high energy events, since number of hits is around 100K.
The voxelization greatly reduces the number of hits, and then size of this matrix.

gondiaz replied:
Probably we dont want voxelization of ionization electrons at the EL anymore, since I think we can use the loop-based implementation for SIPM signal instead the matrix-based. I would left the function here until we decide.

gonzaponte commented on invisible_cities/config/detsim.conf:
```diff
@@ -0,0 +1,32 @@
+files_in = "/Users/gonzalo/Documents/NEXT/detsimprod/data/Kr_full_test.sim.h5"
+file_out = "/Users/gonzalo/Documents/NEXT/detsimprod/out.h5"
```
Not sure what is this file doing in this PR, but in any case, these should be paths like `$ICDIR/database/test_data/blabla.h5` and `/tmp/blabla.h5`

gondiaz replied:
Right, this shouldn't be here

gonzaponte commented on invisible_cities/config/detsim.conf:
```diff
@@ -0,0 +1,32 @@
+run_number = -1
+
+krmap_filename   = "/Users/gonzalo/Documents/NEXT/detsimprod/data/MC_NEW_Kr_map.h5"
+psfsipm_filename = "/Users/gonzalo/Documents/NEXT/detsimprod/data/PSF_SiPM_dst_sum_collapsed.h5"
```
Same comment as above.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+import numpy as np
+
+from typing import List
```
type annotations use `list` rather than `List`. I think the convention is to use `List`.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,140 @@
+from typing import List
+from typing import Tuple
+
+from invisible_cities.cities.detsim_get_psf import binedges_from_bincenters
```
This line means that another PR must go first, otherwise this import will fail.

gonzaponte reviewed (changes requested):
I find the order of the arguments a bit weird. Usually, the prescription is: the data the function depends on, in a heuristically-determined order of importance and then "configuration" parameters. For example, in `generate_ionization_electrons`, `energies` is clearly the main variable while `wi` and `fano_factor` are parameters of the algorithm.

And... we need tests.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,153 @@
+        distributed with mean N and variance (N*F). (Being F the fano_factor).
+        If the variance is such that <1, the electrons are assumed to be poisson distributed.
+    """
+    nes = np.array(energies/wi)
```
no need for `np.array`

gonzaponte replied:
Also, use spaces around `/ `

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,153 @@
+    """
+    nes = np.array(energies/wi)
+    var = nes * fano_factor
+    pois = var<1
```
Inline comment referring to docstring.

gonzaponte replied:
Also, use spaces around `<`

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,153 @@
+    nes = []
+    for t, e in zip(ts, electrons):
+        nes.append(np.count_nonzero(-lifetime * np.log(np.random.uniform(size=e)) > t))
+    return np.array(nes, dtype=int)
```
Despite not being wrong *per se*, this implementation will be slower. But more importantly, I don't think it adds any clarity to the implementation (in my opinion, it takes more effort to read). It also adds an unnecessary np.array call at the end. I am not trying to impose my implementation, but I don't see any need to bring a loop to python when it was already neatly implemented in numpy.
Remove commented-out code, please.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,153 @@
+    var = nes * fano_factor
+    pois = var<1
+    nes[ pois] = np.random.poisson(nes[pois])
+    nes[~pois] = np.random.normal(nes[~pois], np.sqrt(var[~pois]))
```
not sure that the `round` operation was unnecessary. casting to `int` doesn't round, it truncates the decimals

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -87,9 +90,9 @@ def diffuse_electrons(transverse_diffusion   : float,
-    zs = np.repeat(zs, electrons.astype(int))
+    xs = np.repeat(xs, electrons)
+    ys = np.repeat(ys, electrons)
+    zs = np.repeat(zs, electrons)
```
This comment goes into another PR, but you can probably find it faster and better. Please ensure whatever piece of code calling this function is feeding it the right type for the argument `electrons`

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,82 @@
+def test_generate_ionization_electrons(MChits_and_detsim_electron_simulation_params):
+    xs, ys, zs, energies,\
+    wi, fano_factor, lifetime, drift_velocity,\
+    transverse_diffusion, longitudinal_diffusion = MChits_and_detsim_electron_simulation_params
```
This adds a lot of noise. Out convention is to use `_` for the unused arguments. In this case:
`_, _, _, energies, wi, factor, *_ = MChits_and_detsim_electron_simulation_params`. Do this for all the tests.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,82 @@
+    secondary_electrons = generate_ionization_electrons(wi, fano_factor, energies)
+
+    assert len(secondary_electrons)==len(energies)
+    assert np.all(secondary_electrons>=0)
```
Please use spaces around the operators such as `==`, `>=`, `<` etc. We are making an effort to keep consistency across the repository.

Good test. Maybe as a sanity check add another one using `energies = np.zeros(10)` and make sure it produces 0 electrons.

gonzaponte replied:
Another test could be to verify that the output is made of integer values (not the type!)

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,82 @@
+    transverse_diffusion, longitudinal_diffusion = MChits_and_detsim_electron_simulation_params
+
+    secondary_electrons = generate_ionization_electrons(wi, fano_factor, energies)
+    electrons = drift_electrons(lifetime, drift_velocity, zs, secondary_electrons)
```
This test shouldn't use `generate_ionization_electrons`. A test should reduce to a minimum the number of external funcions used. Just make up a sequence for `secondary_electrons` and use it for the test.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,82 @@
+    secondary_electrons = generate_ionization_electrons(wi, fano_factor, energies)
+    electrons = drift_electrons(lifetime, drift_velocity, zs, secondary_electrons)
+
+    assert np.all(electrons>0)
```
I guess it is unlikely, but it is not strictly true that the number of electrons is greater than 0. Make sure that the number of electrons in all hits is large enough for this not to happen and add an inline comment clarifying this.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,82 @@
+    electrons = drift_electrons(lifetime, drift_velocity, zs, secondary_electrons)
+
+    assert np.all(electrons>0)
+    assert np.sum(electrons)<=np.sum(secondary_electrons)
```
How about `np.all(electrons <= secondary_electrons)`? So you ensure that each hit separately doesn't create electrons.

gonzaponte replied:
Another test would be to use infinite and zero lifetime and check that the result is **exactly** what we would expect. And another one to check that the output is made of integer values (again, not the type, the value).

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,82 @@
+    transverse_diffusion, longitudinal_diffusion = MChits_and_detsim_electron_simulation_params
+
+    secondary_electrons = generate_ionization_electrons(wi, fano_factor, energies)
+    electrons = drift_electrons(lifetime, drift_velocity, zs, secondary_electrons)
```
same comment as above. Just make up some values, rather than calculating them using other functions. The reasoning behind this is that if those functions break these tests will too, and it may mislead you to look for the error in the wrong place.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,82 @@
+    electrons = drift_electrons(lifetime, drift_velocity, zs, secondary_electrons)
+    (dxs, dys, dzs) = diffuse_electrons(transverse_diffusion, longitudinal_diffusion,
+                                        xs, ys, zs, electrons)
+    assert len(dxs)==len(dys)==len(dzs)==np.sum(electrons)
```
This test is good. Maybe another one ensuring edge cases like using `z = -10` doesn't yield a hit with `z < 0`?  This would be to verify that the z-clipping is working.

gonzaponte replied:
Another test could also be setting the diffusion values to zero and check that the positions are still the same.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,153 @@
+             dx         : np.ndarray,
+             dy         : np.ndarray,
+             dz         : np.ndarray)\
+             ->Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
```
If the function is staying, it needs tests!

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,82 @@
+    hits_df    = load_mchits_df(nexusMC_filename)
+    event_ids  = hits_df.index.levels[0].values
+    #choose random event
+    evt = np.random.choice(event_ids)
```
Better fix the event number. Otherwise, if an error happens, it is more difficult to reproduce the result.

gonzaponte reviewed (changes requested):
> General comments:
> The files are in the wrong place. cities should contain cities and components of source type (in components.py). These are general functions that belong either in a folder dedicated to detector simulation or in reco maybe. I think they should be moved.

This request by @andLaing should also be addressed (probably right before merging, to avoid confusion and help the review process).

There are other issues raised previously (more physics-related) that I am not sure are closed.

jacg commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -61,7 +60,7 @@ def test_generate_ionization_electrons(MChits_and_detsim_electron_simulation_par
     _, _, zs, energies,\
     wi, fano_factor, lifetime, drift_velocity,\
-    _, _ = MChits_and_detsim_electron_simulation_params
+    *_ = MChits_and_detsim_electron_simulation_params
```
Please indent continued lines. (Not only on line 63, but also on 52 ... and anywhere else they might be.)

Without the indentation, the continuation looks like a line in its own right, so I wasted far too much time staring at 

```python
*_ = just_some_identifier
```
and wondering what the point was. Only when noticing the `\` at the end of the *previous* line did it make any sense. Indentation of the continuation would have made it obvious that it's part of something that precedes it.

Generally speaking, try to avoid continued lines: they are fragile and confusing. Frequently you can replace them with brackets, but not in this case.

In this specific case, you are unpacking something that contains over 8 components and relying on getting the order right. To me this is begging for a `namedtuple` or some other way of naming these components.

jacg commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -56,6 +56,13 @@ def test_generate_ionization_electrons(MChits_and_detsim_electron_simulation_par
     assert len(secondary_electrons) == len(energies)
     assert np.all(secondary_electrons >= 0)
 
+    isinteger = lambda value: int(value) == value
```
Anything wrong with `float.is_integer`?

gondiaz replied:
This works if the values were `float`. `generate_ionization_electrons` and `drift_electrons` return `numpy.ndarray` of type `numpy.int64` values.

jacg replied:
Apologies if I'm being thick, but if `generate_ionization_electrons` returns an array of `int64`s, then I'm struggling to see the point of checking whether they are all integers. (Maybe I've been spending too much time with static type systems lately?)

(You mention `drift_electrons` in response to a comment about `isinteger` but I don't see how the two are related.)

Also, whilst we're here, the fact that line 63 is character-for-character identical to line 54 is a bit confusing. I appreciate that the energies are different, but using the same name in the same scope for a different set of energies promotes confusion.

Line 62 onwards seems to be completely independent of what precedes it (besides using the same fixture), which suggests that it belongs in a separate test.

gondiaz replied:
For the integers I was following the comments: https://github.com/nextic/IC/pull/695#discussion_r467050629, https://github.com/nextic/IC/pull/695#discussion_r467050093. Maybe I have not understand it properly, @gonzaponte.

I do not know if it worths to split the tests for null `energies` in `generate_ionization_electrons`; for null and infinite `lifetime` in `drift_electrons` and for null diffusions and negative `zs` in `diffuse_electrons`...

jacg replied:
I don't understand the comments you link: If the type is an integer type, then the values are guaranteed to be integers! If the type is `float` then you should use `float.is_integer` rather than baking your own. The only situation in which it might make sense is if *sometimes* the values are `float`s and *sometimes* `int`s ... but you claim that they are *always* `int64`, so I'm lost.

As for splitting the tests: tests should be as orthogonal, independent and short as is *reasonably* possible (I'll be happy to explain this at great length, if you wish, but this is not the place for it). As ever, don't take this as dogma, but as a guiding principle. These two bits of the test are idependent. Is there any reason why they should be grouped under the same test? I don't see the reason, but I'd be happy for it to be pointed out to me.

gonzaponte replied:
Sorry, I'm still catching up with this. I haven't been able to keep an eye on this yesterday.

> Maybe I have not understand it properly, @gonzaponte.

No, I think you understood correctly. I meant that we want to check the value, not the type.

> If the type is an integer type, then the values are guaranteed to be integers! 

Indeed. But the function returning an int-array type is just an implementation choice (more intuitive for the human mind). Everything should work the same whether we have an int-type array or a float-type array holding integers only (I think). I suggested value-based tests following the duck typing paradigm. I want to make sure we don't return non-integer values, not non-integer types. Does it sound right?

As for the splitting the tests, yes, please. Make an individual test for each feature/behavior as much as possible.

jacg replied:
So the value of the `isinteger` check is protection against the return type of `generate_ionization_electrons` changing to some kind of `float` which would then admit non-integer values, which might then actually appear in the array?

Hmm, OK ...

> I suggested value-based tests following the duck typing paradigm.

I wonder whether that's taking it too far. Just because Duck Typing works well in Python in many cases, and we should use it to our advantage when we can, that doesn't mean that we should religiously try to force it in everywhere: in some cases Duck Typing is simply not the best option. If it is nonsensical for these values to be non-integers, I think that using the (non-Duck) type system to say so, is probably more appropriate.

How you express this idea with sufficient generality in Python's contrete type system is another question. There's `numpy.integer` but unfortunately the builtin `int` is not related to it.

[As a fun aside, let's nod to Purescript's type system, which would allow us to specify concrete type constraints that nail down the [exact algebraic properties](https://a-guide-to-the-purescript-numeric-hierarchy.readthedocs.io/en/latest/introduction.html) that these "numbers" should have! In other words, it's a static-formal-duck-type-system[*] which allows you to say exactly what you want with remarkable precision.

[*] By formal-duck I mean that it not only has to quack, but the quack actually has to have the intended meaning.]

gonzaponte replied:
Ok, so you suggest we stick to return integer values and check for the type in the tests?

jacg replied:
Don't take my word as gospel, especially in this case where I've only scratched the surface (so, perhaps if I look at more context, my opinion will change), but a type check looks reasonable to me at the moment.

The problem is generality. How about something like `is{instance/subclass}(..., (np.integer, int))`?

gonzaponte replied:
From your previous explanation, it sounds reasonable. We are definitely working with something that doesn't make sense in non-integer values (you cannot have 3.5 electrons), so it makes sense to restrict the type. Now, within the type `integer` we don't care which one it is as long as it does not limit the values (we don't want `np.int8` or `np.int16`). That aside, should we do something like `"int" in array.dtype.name`?

jacg replied:
> `"int" in array.dtype.name`

Bleargh!

It's most unfortunate that the use of strings to represent more meaningful and structured information, has gained such popularity in Python culture: it's quite an abominable practice. I see the attraction here, as it lets you write something that isn't obviously incorrect at first blush in not too many characters, but ... we don't care about a relationship between sequences of characters, we care about types! And types are first-class objects in Python!

Can I encourage you to eschew text-pattern-matching and tend towards using objects, and their meaningful relationships and/or identities?

How about `issubclass(arr.dtype.type, (np.integer, int))`?

gonzaponte replied:
> Bleargh!

I know, I know.. But I thought there was no way to include `uint`s in the equation otherwise. Fortunately 

> How about issubclass(arr.dtype.type, (np.integer, int))?

works, so let's go for it.

gondiaz replied:
> Indeed. But the function returning an int-array type is just an implementation choice (more intuitive for the human mind). Everything should work the same whether we have an int-type array or a float-type array holding integers only (I think). I suggested value-based tests following the duck typing paradigm. I want to make sure we don't return non-integer values, not non-integer types. Does it sound right?

No, in any case they should be int-types since in `drift_electrons` they enter as a `size` parameter and in `diffuse_electrons` in `np.repeat` as the repeats. So a float-type will not work even if it is an integer-value.

gonzaponte replied:
Correct. I had tried with a float value in `np.repeat` and it worked, but not with a `np.ndarray` containing floats.

jacg commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -56,6 +56,13 @@ def test_generate_ionization_electrons(MChits_and_detsim_electron_simulation_par
     assert np.all(secondary_electrons >= 0)
 
+    isinteger = lambda value: int(value) == value
+    assert np.all(list(map(isinteger, secondary_electrons)))
```
Is `np.all(list(map(...)))` really better than `all(map(...))`?

jacg commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -48,35 +48,36 @@ def MChits_and_detsim_electron_simulation_params(nexusMC_filename):
-    secondary_electrons = generate_ionization_electrons(wi, fano_factor, energies)
-    electrons = drift_electrons(lifetime, drift_velocity, zs, secondary_electrons)
+    secondary_electrons = generate_ionization_electrons(energies, wi, fano_factor)
+    electrons = drift_electrons(zs, secondary_electrons, lifetime, drift_velocity)
```
As I alluded elsewhere, all this ignoring and reordering is begging for named components.

jacg commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -81,7 +81,7 @@ def test_drift_electrons(MChits_and_detsim_electron_simulation_params):
     # test is integer
     isinteger = lambda value: int(value) == value
-    assert np.all(list(map(isinteger, drifted_electrons)))
+    assert np.all(map(isinteger, drifted_electrons))
```
```python
In [1]: import numpy as np                                                                                                                                                                                                                                                        

In [2]: all(map(lambda x:x, range(10**6)))                                                                                                                                                                                                                                        
Out[2]: False

In [3]: np.all(map(lambda x:x, range(10**6)))                                                                                                                                                                                                                                     
Out[3]: <map at 0x7fa3a530f050>

In [4]: bool(_2), bool(_3)                                                                                                                                                                                                                                                        
Out[4]: (False, True)
```

jacg commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -42,14 +44,21 @@ def MChits_and_detsim_electron_simulation_params(nexusMC_filename):
+
+    typename = "MChits_and_detsim_params"
+    field_name = ["xs", "ys", "zs", "energies", "wi", "fano_factor", "lifetime",
+                  "drift_velocity", "transverse_diffusion", "longitudinal_diffusion"]
```
Lists of strings are very annoying to type, and quite noisy to read, so avoid doing it ... especially when Python is working hard to make your life easy for you:

```python
In [1]: from collections import namedtuple                                                                                                                                                                                                                                        

In [2]: names = "this that other"                                                                                                                                                                                                                                                 

In [3]: XXX = namedtuple("XXX", names)                                                                                                                                                                                                                                            

In [4]: XXX(1,2,3)                                                                                                                                                                                                                                                                
Out[4]: XXX(this=1, that=2, other=3)
```

jacg replied:
And if the line gets too long for your liking, triple-quoted strings will work just fine:

```python
In [1]: from collections import namedtuple                                                                                                                                                                                                                                        

In [2]: lots_of_names = """one two three 
   ...:    four five six"""                                                                                                                                                                                                                                                       

In [3]: namedtuple("X", lots_of_names)(1,2,3,4,5,6)                                                                                                                                                                                                                               
Out[3]: X(one=1, two=2, three=3, four=4, five=5, six=6)
```

jacg commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -42,14 +44,21 @@ def MChits_and_detsim_electron_simulation_params(nexusMC_filename):
+
+    nt = namedtuple(typename, field_name)
+    return nt(xs, ys, zs, energies, wi, fano_factor, lifetime,
+              drift_velocity, transverse_diffusion, longitudinal_diffusion)
```
To increase safety, you could consider keyword args: `return nt(xs=xs, ys=ys ...)`. Tedious, but safe.

To alleviate the tedium you could do something like

```python

In [1]: from collections import namedtuple                                                                                                                                                                                                                                        

In [2]: from operator import itemgetter                                                                                                                                                                                                                                           

In [3]: one, two, three = 10,20,30    # Replace with the actual variables you use                                                                                                                                                                                                                                  

In [4]: names = 'one two three'                                                                                                                                                                                                                                                   

In [5]: XXX = namedtuple('XXX', names)                                                                                                                                                                                                                                            

In [6]: XXX(*itemgetter(*XXX._fields)(locals()))                                                                                                                                                                                                                                  
Out[6]: XXX(one=10, two=20, three=30)
```

which is more cryptic, but is quite DRY and thus makes it difficult to mess up the order, now or in the future.

jacg commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -42,14 +44,21 @@ def MChits_and_detsim_electron_simulation_params(nexusMC_filename):
+
+    energies    = MChits_and_detsim_params.energies
+    wi          = MChits_and_detsim_params.wi
+    fano_factor = MChits_and_detsim_params.fano_factor
```
That's fine ... but in some cases you might consider an approach along these lines:

```python
f = MChits_and_detsim_params
secondary_electrons = generate_ionization_electrons(f.energies, f.wi, f.fano_factor)
```

Here I picked `f` for **F**ixture, but something else might be more appropriate depending on context.

jacg commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -93,7 +93,7 @@ def test_drift_electrons(MChits_and_detsim_params):
     # test is integer
     isinteger = lambda value: int(value) == value
-    assert np.all(map(isinteger, drifted_electrons))
+    assert all(map(isinteger, drifted_electrons))
```
I trust that this commit will end up being squashed into the original.

jacg commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,137 @@
+    field_name = """xs ys zs energies wi fano_factor lifetime drift_velocity
+                    transverse_diffusion longitudinal_diffusion"""
+
+    nt = namedtuple(typename, field_name)
```
Why not use keyword arguments in `namedtuple`?

```suggestion
    nt = namedtuple(typename = "MChits_and_detsim_params", 
                    field_names = """xs ys zs energies wi fano_factor lifetime
                       drift_velocity transverse_diffusion longitudinal_diffusion""")
```
You may prefer some other layout of `field_names`.

jacg commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -67,8 +65,13 @@ def test_generate_ionization_electrons(MChits_and_detsim_params):
     # test is integer
-    isinteger = lambda value: int(value) == value
-    assert all(map(isinteger, secondary_electrons))
+    assert issubclass(secondary_electrons.dtype.type, np.integer)
```
This will fail if the dtype is `int`, which is why `np.integer` should be replaced with `(np.integer, int)`.

jacg commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -119,8 +119,8 @@ def test_drift_electrons_extreme_lifetimes(MChits_and_detsim_params):
 def test_diffuse_electrons(MChits_and_detsim_params):
 
-    xs                     = MChits_and_detsim_params.zs
-    ys                     = MChits_and_detsim_params.zs
```
This sort of bug is exactly why I prefer the style I suggested earlier:

```python
f = MChits_and_detsim_params
use_fixture(f.xs, f.ys)
```

There are two main problems with the style you have chosen

1. It adds a bunch of pretty pointless lines to the code
2. The signal-to-noise ratio of those lines is veeeery low (which is why you didn't spot the bug immediately).

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,152 @@
+from typing import List
+from typing import Tuple
+
+# from invisible_cities.cities.detsim_get_psf import binedges_from_bincenters
```
Remove commented-out code.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,152 @@
+def drift_electrons(zs             : np.ndarray,
+                    electrons      : np.ndarray,
+                    lifetime       : float = 12.,
+                    drift_velocity : float = 1.) -> np.ndarray:
```
```suggestion
                    lifetime       : float = 12 * units.ms,
                    drift_velocity : float = 1 * units.mm / units.mus) -> np.ndarray:
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,152 @@
+#######################################
+def generate_ionization_electrons(energies    : np.ndarray,
+                                  wi          : float = 22.,
+                                  fano_factor : float = 0.1) -> np.ndarray:
```
```suggestion
def generate_ionization_electrons(energies    : np.ndarray,
                                  wi          : float = 22 * units.eV,
                                  fano_factor : float = 0.1) -> np.ndarray:
```
Are these values correct? I think 22.4 and 0.15 are the correct ones.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,152 @@
+                      zs                     : np.ndarray,
+                      electrons              : np.ndarray,
+                      transverse_diffusion   : float = 1.,
+                      longitudinal_diffusion : float = 0.2)\
```
```suggestion
                      transverse_diffusion   : float = 1 * units.mm / np.sqrt(units.cm),
                      longitudinal_diffusion : float = 0.2  * units.mm / np.sqrt(units.cm))\
```
Again, the most recent numbers are 1.1 and 0.3, I think.

gondiaz replied:
I was using 1 and 0.2 in the last productions

gonzaponte replied:
@paolafer Which values should we set as default?

paolafer replied:
The current values of NEXT100, extrapolated from the measurements made in NEW, are 1 and 0.2. I don't have a strong opinion on which values the default should have: these functions are general, and not meant to simulate NEXT100 in this conditions of temperature and pressure only.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,152 @@
+def diffuse_electrons(xs                     : np.ndarray,
+                      ys                     : np.ndarray,
+                      zs                     : np.ndarray,
+                      electrons              : np.ndarray,
```
can we rename this to `n_electrons`, `n_ie`, or something like that? The idea is to make it clear the it contains _a number of_ electrons

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,152 @@
+#         return dx, dy, dz, nes
+#
+#     else:
+#         return dx, dy, dz, np.ones(dx.shape)
```
Remove commented-out code. Please try to avoid leaving commented code around.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,142 @@
+
+@fixture(params = ["Kr83_nexus_v5_03_00_ACTIVE_7bar_3evts.MCRD.h5"])
+def nexusMC_filename(request, ICDATADIR):
+    return os.path.join(ICDATADIR, request.param)
```
Do you plan to add other filenames? If not the parametrization of the fixture is unnecessary.

gondiaz replied:
This fixture is not really necessary (it is already defined in `conftest.py`, `krypton_MCRD_file`). I just copied it here for clarity, I remove it now.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,142 @@
+    return os.path.join(ICDATADIR, request.param)
+
+
+@fixture
```
The default scope is `function`. I guess you don't want to execute the code of this fixture every time a test needs it, right? If so:
```suggestion
@fixture(scope="session")
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,142 @@
+
+    nt = namedtuple(typename = "MChits_and_detsim_params",
+                    field_names = """xs ys zs energies wi fano_factor lifetime
+                       drift_velocity transverse_diffusion longitudinal_diffusion""")
```
```suggestion
    nt = namedtuple(typename    = "MChits_and_detsim_params",
                    field_names = """xs ys zs energies
                                     wi fano_factor
                                     lifetime drift_velocity
                                     transverse_diffusion longitudinal_diffusion""")
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,142 @@
+    f = MChits_and_detsim_params
+
+    n_ie = np.mean(f.energies / f.wi)
+    secondary_electrons = np.clip(np.random.normal(n_ie, n_ie**0.5, size=len(f.zs)), 0, None).astype(int)
```
This is ok, but for future references, you don't need to recreate the overall algorithm to make a test. In this case, you just want to ensure some basic behavior of a function. it would have been as good, if not better, to just use a custom array of integers without necessarily being consistent with the physics behind it. That will make the test simpler and easier to understand.

gondiaz replied:
Right, I used this to make sure the length of `secondary_electrons` (now n_electrons) is the same as `f.energies` an to be an array of integers. I think it is clearer in this case to be this way than building another array like a `np.arange(len(f.energies))`

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,142 @@
+    sel = np.random.choice(np.arange(0, len(f.zs)), size=np.random.randint(1, len(f.zs)))
+    f.zs[sel] = -1
+    (dxs, dys, dzs) = diffuse_electrons(f.xs, f.ys, f.zs, electrons, f.transverse_diffusion, f.longitudinal_diffusion)
+    assert np.all(dzs >= 0)
```
Separate test

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,142 @@
+    n_ie = np.mean(f.energies / f.wi)
+    electrons = np.clip(np.random.normal(n_ie, n_ie**0.5, size=len(f.zs)), 0, None).astype(int)
+
+    (dxs, dys, dzs) = diffuse_electrons(f.xs, f.ys, f.zs, electrons, f.transverse_diffusion, f.longitudinal_diffusion)
```
The parentheses on the left-hand side are not needed, but you can keep them if you think it helps visually.

gondiaz replied:
I think it helps, for example to distinguish the case when we do not want all the returns, as `(dxs, dys, _)` in the last test. Also emphasizes the line where the function call is being done.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,105 @@
+    sqrtz = zs ** 0.5
+    dxs  = np.random.normal(xs, sqrtz * transverse_diffusion)
+    dys  = np.random.normal(ys, sqrtz * transverse_diffusion)
+    dzs  = np.random.normal(zs, sqrtz * longitudinal_diffusion)
```
```suggestion
    dxs  = np.random.normal(xs, sqrtz *   transverse_diffusion)
    dys  = np.random.normal(ys, sqrtz *   transverse_diffusion)
    dzs  = np.random.normal(zs, sqrtz * longitudinal_diffusion)
```

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,148 @@
+import os
+import numpy  as np
+import tables as tb
+import pandas as pd
```
`os`, `tb` and `pd` not used. Remove imports

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,148 @@
+
+    #hits
+    hits_df    = load_mchits_df(krypton_MCRD_file)
+    event_ids  = hits_df.index.levels[0].values
```
`event_ids` not used.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons.py:
```diff
@@ -0,0 +1,105 @@
+import numpy as np
+
+from typing import List
```
`List` not used.

gonzaponte commented on invisible_cities/cities/detsim_simulate_electrons_test.py:
```diff
@@ -0,0 +1,148 @@
+
+    # test z<0 clip
+    zs = np.copy(f.zs)
+    sel = np.random.choice(np.arange(0, len(zs)), size=np.random.randint(1, len(zs)))
```
I think this line overcomplicates the test and adds randomness, which avoids reproducibility. Just make up a set of indices and use it.

I understand the point of randomness to avoid magic numbers, and that's in part why we use hypothesis, but there is a trade-off. Your test will be more robust if you don't cherry-pick numbers yourself, but it also adds a layer of complexity. Is it worth it for this kind of simple test? Probably not. In this specific case, there is no good reason why the test would give a false positive for a specific set of indices unless the function is coded with malice.

jacg replied:
> there is no good reason why the test would give a false positive for a specific set of indices unless the function is coded with malice.

Te-he, you'd be amazed at the bugs that property-based testing has found over the years, bugs that you would imagine are only possible through malice, until they actually appear in production code through innocent error. That's one of the big features of property-based testing: it has a habit of finding failing test cases that humans assume would never need testing because it would be absurd for them to fail given the cases that the humans have already checked. 

However, you do make a valid point: you have to find a reasonable balance between effort invested, clarity of code and benefits gained, rather than religiously testing absolutely everything to the max.

If you *are* going to introduce randomness into tests, please try to do it with hypothesis, rather than ad-hoc calls to random number generators. Once again, make some costs/benefits analysis rather than taking this as religious dogma.

gonzaponte reviewed (changes requested):
Just some cleanup and very minor changes. After that, this will be ready for approval.

gonzaponte commented on invisible_cities/detsim/simulate_electrons.py:
```diff
@@ -0,0 +1,102 @@
+
+from typing import Tuple
+
+import invisible_cities.core.system_of_units as units
```
```suggestion
import .. core.system_of_units as units
```

gonzaponte commented on invisible_cities/detsim/simulate_electrons_test.py:
```diff
@@ -0,0 +1,144 @@
+
+from invisible_cities.cities.detsim_simulate_electrons import generate_ionization_electrons
+from invisible_cities.cities.detsim_simulate_electrons import drift_electrons
+from invisible_cities.cities.detsim_simulate_electrons import diffuse_electrons
```
```suggestion
import .. core.system_of_units as units

from ..io.mcinfo_io import load_mchits_df

from . simulate_electrons import generate_ionization_electrons
from . simulate_electrons import drift_electrons
from . simulate_electrons import diffuse_electrons
```
(notice that the `detsim_` prefix is also removed)

gonzaponte reviewed (commented):
Sorry, I forgot to mention the relative imports.

gondiaz committed:
Add electron simulation module

This module contains the functions to generate, diffuse and drift
ionization electrons

gondiaz committed:
Add tests

gonzaponte reviewed (approved):
This PR implements all the functions needed to transform energy deposits from nexus into ionization electrons at the EL gap.  The code is well written, documented, and tested.
Good job.
